### ISSUE TITLE
Overlapping display of the upper vowel and the tone character in Thai Language

### ISSUE THREAD

**Issue by nthana** [godotengine/godot#28371]
[Godot v. 3.1 stable]
[Tested on: Godot on MacOS]

In Thai language, a consonant character may be followed by an upper vowel, a tone character, or both. These characters must be displayed above the consonant. When these characters appear simultaneously, the tone character's vertical position should be moved to above the upper vowel character, preventing the overlap.

Currently, Godot can display upper vowels and tone characters correctly. However, when both occurs, they are incorrectly displayed in run-time (overlapping occur), but correctly displayed on the label-text editor in design-time. 

The test project: https://github.com/nthana/ThaiFont 

In label-text editor (correctly displayed):
<img width="529" alt="In label-text editor (correctly displayed)" src="https://user-images.githubusercontent.com/44568630/56631701-a87b2300-6680-11e9-8a12-38f553f8845b.png">

On runtime (incorrectly displayed):
<img width="586" alt="On runtime (incorrectly displayed)" src="https://user-images.githubusercontent.com/44568630/56632088-56d39800-6682-11e9-8d2c-3499808c4d82.png">

Thanks.

**Comment by akien-mga** [comment]
The difference between the in-game display and the editor display is due to the font `Trirong-Medium.ttf`. If you use the same font as editor font, you'll see the problem appear in the Inspector too.

The font itself is fine though, here's how it looks like in LibreOffice:
![Screenshot_20190424_123404](https://user-images.githubusercontent.com/4701338/56653121-4117a000-668d-11e9-867c-c22194d62cd9.png)

So there's an issue in how we/freetype handles this font.

**Comment by nthana** [comment]
@akien-mga Thank you. I'm grateful for the response. I'm totally agree with you. The font is fine, because I use this font with other applications. I also tried with other fonts, and the problem still occur. 

I couldn't help on C++ coding because my knowledge on C++ & the Godot source code is limited. 
But I am very glad to help on how the algorithm should be. With my Thai language knowledge, I will try to describe the algorithm on my next post. The algorithm is not too complex.

p.s. I'm native Thai and I've currently strived to promote Godot Engine usage in Thailand.

**Comment by nthana** [comment]
The below picture shows all of Thai characters in Unicode. 
[adapt from source: https://jrgraphix.net/r/Unicode/0E00-0E7F ]

<img width="842" alt="Thai Unicode" src="https://user-images.githubusercontent.com/44568630/56657163-8f469680-66c1-11e9-9030-3bb56eaedabe.png">

Upper Vowels : shown in Blue frames.
Tone characters : shown in Red frame.  

The algorithm is:
- whenever an "Upper Vowel" is followed by a "Tone Character", The tone character should be vertically raised to prevent overlapping.

Thanks.

**Comment by starry-abyss** [comment]
> The algorithm is:
>
>  whenever an "upper vowel" is followed by a "tone character", The tone character should be vertically >raised to prevent overlapping.

Not a language guy, but I guess if there's an algorithm involved, one might need this feature https://github.com/godotengine/godot/pull/21791

**Comment by nthana** [comment]
@starry-abyss Wow! From [the screen shot of #21791](https://github.com/godotengine/godot/pull/21791#issuecomment-425040469), it displays Thai language perfectly. 

So, may be, this issue is the sub-problem of that general implementation in [#21791](https://github.com/godotengine/godot/pull/21791)

**Comment by akien-mga** [comment]
Since it works with the default editor font though (`NotoSansThaiUI_Regular.ttf`), it seems like we can already support the features needed for Thai even without advanced font shaping needed for e.g. Arabic.

Here's how Label looks like in game with `NotoSansThaiUI_Regular.ttf` (+ `NotoSansUI_Regular.ttf` as fallback to have the latin glyphs):
![Screenshot_20190425_092418](https://user-images.githubusercontent.com/4701338/56717009-fdc83a80-673b-11e9-83c0-4c8b9b3a2b30.png)

The diacritics are not shown on the two middle lines, which is weird as they are actually visible in the Inspector.

**Comment by akien-mga** [comment]
With both fonts, I can get the diacritics to show on the middle lines by changing the `extra_spacing/char` property of the DynamicFontData.

E.g. here with 1:
![Screenshot_20190425_094305](https://user-images.githubusercontent.com/4701338/56718206-a11a4f00-673e-11e9-8fda-ad7aa7a90415.png)

And with -3:
![Screenshot_20190425_094418](https://user-images.githubusercontent.com/4701338/56718250-b68f7900-673e-11e9-9750-251a62c17fcb.png)

So yeah, there's definitely something fishy in the way DynamicFont handles those.

**Comment by nthana** [comment]
@akien-mga Thank you so much.

I think the reason that sometimes the diacritics are not shown is because there is no meaning in Thai when diacritics appear without preceded consonants. So, many font rendering engines will not render a diacritic alone. This also happen in many softwares such as Microsoft Office, and it is totally acceptable because a diacritic will never be used alone, in real-life sentence.

I think the reason that the NotoSansThaiUI_Regular.ttf is displayed properly is because its diacritics' positions are designed to always be at top most level to avoid overlapping with any upper vowels. However, many new Thai fonts place diacritics and vowels at the same vertical positions (because of aesthetic reason). These fonts will need additional placement logics like in [#21791](https://github.com/godotengine/godot/pull/21791).

In conclusion, at this time, if Thai language display is needed, the NotoSansThaiUI_Regular.ttf font should be used. However, for other Thai fonts, we should wait for the complex glyph positioning in #21791](https://github.com/godotengine/godot/pull/21791) to be completely implemented.

Thanks again :)

**Comment by nthana** [comment]
Just for information. I have additionally tested some Thai fonts and found that:

Some Thai Fonts that correctly displayed by Godot 3.1:
Arial Unicode.ttf
Ayuthaya.ttf
Krungthep.ttf
NotoSansThaiUI-Regular.ttf
Sathu.ttf
Silom.ttf
So, the worked fonts are much more than I previously understand.

Some Thai Fonts that incorrectly displayed by Godot 3.1 (Overlapping Occurs):
Kanit-Regular.ttf
Pridi-Regular.ttf
Prompt-Regular.ttf
Trirong-Medium.ttf

**Comment by bruvzg** [comment]
Should be fixed by CTL implementation (#41100):
<img width="593" alt="Screenshot 2020-11-10 at 10 54 40" src="https://user-images.githubusercontent.com/7645683/98651699-4148c080-2343-11eb-88e2-b597fd9cfaf6.png">
